### PR TITLE
Loosen type annotations

### PR DESCRIPTION
**Pull request by devmotion**
This is required for the change in https://github.com/JuliaNLSolvers/NLsolve.jl/pull/117 that allows arbitrary `AbstractArray`.

### CODE REVIEW

File: src/objective_types.jl [hunk 1]
```diff
@@ -3,41 +3,41 @@ real_to_complex(d::AbstractObjective, x) = iscomplex(d) ? real_to_complex(x) : x
...
 
 # Used for objectives and solvers where no gradient is available/exists
-mutable struct NonDifferentiable{T,Tcplx} <: AbstractObjective where {T<:Real,
+mutable struct NonDifferentiable{T,A<:AbstractArray{T},Tcplx} <: AbstractObjective where {T<:Real,
                                                             Tcplx<:Union{Val{true},Val{false}}  #if true, must convert back on every f call
                                                             }
     f
     f_x::T
-    last_x_f::Array{T}
+    last_x_f::A
```
Comment (anriseth) [inline src/objective_types.jl]: What's the advantage of doing this, instead of `last_x_f::AbstractArray{T}` ?

Comment (devmotion) [reply src/objective_types.jl]: The compiler can infer the type of `last_x_f` and generate high performance code. In the documentation, at https://docs.julialang.org/en/stable/manual/performance-tips/#Avoid-fields-with-abstract-type-1 and https://docs.julialang.org/en/stable/manual/performance-tips/#Avoid-fields-with-abstract-containers-1, it is explained in more detail why one should not use fields with abstract types.

Comment (anriseth) [reply src/objective_types.jl]: Ah, thanks! That's probably something I've messed up in a lot of my code the last months.

**Comment by pkofod** [comment]
lgtm, but did you make sure to test Optim, NLsolve and LsqFit with this? Should regress, but sometimes subtle things can surprise you.

**Comment by devmotion** [comment]
I tested it against master branches of Optim, NLsolve, and LsqFit, and compared it with NLSolversBase master (I had to update https://github.com/JuliaNLSolvers/Optim.jl/blob/master/test/general/counter.jl, as indicated in the comments). All tests passed and I could not discover any regression.

**Comment by anriseth** [comment]
Looks good to me. Let's merge @pkofod ?

**Comment by pkofod** [comment]
:+1:

**Comment by anriseth** [comment]
And tag, so that tests on JuliaNLSolvers/NLsolve.jl#117 are satisfied? (You are probably the only one with tagging rights @pkofod :p )

**Pull request merged**